### ISSUE TITLE
Issue with Anaconda package built against openmpi

### ISSUE THREAD

**Issue by kpwelsh** [open-mpi/ompi#11791]
## Background information
I am building an Anaconda distribution of some C++ software parallelized with MPI. I am building two packages independently that both link against libmpi and one of the packages depends on the other.

### What version of Open MPI are you using? (e.g., v3.0.5, v4.0.2, git branch name and hash, etc.)
I am building against the anaconda package openmpi=4.1.5=h414af15_101. 

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)

Installed through "mamba install openmpi." I don't have any versions/builds pinned or anything.

### Please describe the system on which you are running

* Operating system/version: Windows 10 Enterprise running Ubuntu-20.04 in WSL1
* Computer hardware: Intel(R) Xeon(R) W-11855M

-----------------------------

## Details of the problem

I can successfully build and run the applications if I:
1. Build them against my system libraries.
2. Create a new conda environment, install the dependencies (including openmpi) and then build the downstream package from source, linking against the conda environment libraries.

However, if I put it in a conda package, build it and install it into a new environment, I receive this cryptic error:

```shell
[<hostname>:09109] opal_ifinit: socket() failed with errno=2
[<hostname>:09109] pmix_ifinit: socket() failed with errno=2
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] pmix:create_socket: socket() failed: No such file or directory (2)
[<hostname>:09109] OPAL ERROR: Unreachable in file pmix3x_client.c at line 112
*** An error occurred in MPI_Init
*** on a NULL communicator
*** MPI_ERRORS_ARE_FATAL (processes in this communicator will now abort,
***    and potentially your MPI job)
[<hostname>:09109] Local abort before MPI_INIT completed completed successfully, but am not able to aggregate error messages, and not able to guarantee that all other processes were killed!
```

What is most confusing to me is that this error sometimes goes away. If I have a single anaconda environment, without openmpi installed in the host system, this error persists indefinitely. However, if I have an _additional_ anaconda environment with openmpi installed, this error stops occurring ~5-10 minutes after installing my package without doing anything. (Note, I am having a hard time determining what is superstition here vs actually relevant).

Also I can install this on our cluster and it seems to run fine without issues. 

Does anyone have any ideas on why this might be happening? I am a bit new to the details of openmpi, so I am quite lost when I see something like "OPAL ERROR".

**Comment by rhc54** [comment]
I take it you start the application using OMPI's `mpirun` - yes? If so, it is `mpirun` that is creating the socket that the client is trying to attach to. The connection info is provided in the client's environment, not a rendezvous file. It sounds like `mpirun` is unable to create the socket, or else the client cannot connect to it. I know nothing about Anaconda, so I cannot help you there.

**Comment by jsquyres** [comment]
I'm afraid I know nothing about Anaconda, either.

Would you happen to have multiple different versions of the PMIx library installed, perchance (e.g., inside your conda environment and outside your conda environment), and somehow you're inadvertently mixing versions?  E.g., compile+build PMIx with version X, but then run with version Y?

**Comment by jsquyres** [comment]
One additional question: what version of PMIx are you using with Open MPI v4.1.5rc2?

Also, looking at your error message closer, it looks like the `socket()` system call is failing across the board -- both in PMIx and outside of PMIx (e.g., in `opal_ifinit()`).  That's... odd.   E.g., all `opal_ifinit()` is trying to do at that point is open a datagram socker.  Is your conda environment doing anything container-ish, perchance?  E.g., setting up network or filesystem namespaces, or even running in a full-fledged container?

**Comment by kpwelsh** [comment]
@rhc54 Yes, I am running it with mpirun. Thanks for the clarification on the connection semantics there. 

@jsquyres 
Multiple version of things across multiple environments is a likely culprit, which is why I think I have ruled that out. I am not an expert in general containerization, but I can provide more information on how thinks are built/linked, and I am fairly sure my conda environment isn't doing anything out of the ordinary (except being used to host native C++ projects in the first place).

* There is no system-wide MPI installation. libmpi only exists in environment directories.
* openmpi is obtained only through the conda package -- and I have checked the same build/version is installed at package build-time and package install time -- and in this case if there were different library SONAMEs it wouldn't even run.
* `which mpirun` finds "<conda-env>/bin/mpirun"
* `which <executable>` finds "<conda-env>/bin/<executable"
* The exact command that yields the error is "mpirun -np 2 <executable>"
* `ldd `which <executable>`` shows its linking to the correct conda environment's libmpi.so.40
* If I restart WSL and reactivate the environment it seems to remedy the issue. 

There is one set of libpmi.so.1/libpmi2.so.1 installed in my system, but `ldd .../libmpi.so.40` doesn't appear to link against it. My understanding was that openmpi was using its internal pmix, but my understanding here is superficial.

Also If there are any environment variables that might be relevant here, I am happy to grep for them, but anaconda functions by setting a huge number of environment variables, so they are a bit of a mess.

**Comment by kpwelsh** [comment]
Also, more recently I tested the package on another machine (HPC cluster) and it appears to work just fine, even when linking against the systems provided libmpi.

**Comment by jsquyres** [comment]
Ah: I missed that this is running in WSL (specifically, WSL1).

These things:

* The issue is intermittent
* Restarting WSL fixes the issue (at least for a while)
* The failure point is when Open MPI is just trying to open a datagram socket (not a filesystem socket)

make me think that the underlying OS/kernel/environment is at fault here.

I don't know a whole lot about WSL, but I know there were giant improvements with WSL2.  Can you try upgrading to WSL2?  We haven't done a lot of (any?) testing with Open MPI on WSL.

**Comment by kpwelsh** [comment]
I would be surprised if this didn't have something to do with WSL. Unfortunately, the only reason I am using WSL 1 is that my workplace uses Symantec for security and it doesn't work correctly with WSL2.

I don't mind blaming it on the shortcomings of WSL1, since "upgrade to WSL2" is a typically reasonable suggestion. What's bugging me is that issue only occurs when trying to use the package after installing it via Anaconda. I have not seen this issue when running the code after compiling -- only after placing the executables/libraries in a new conda environment. 


Are there any tools I could look at to try to debug the socket creation? It looks like PMIx has some hooks for printing the socket URI its trying to connect to, but I am not sure how to set those flags. I have tried to debug linux sockets at the OS level before to no avail, but perhaps someone knows more tricks than me :).

**Comment by jsquyres** [comment]
This is unfortunately where my lack of Anaconda knowledge comes into play: I don't know what Anaconda does in its environments.

I didn't look closely at the PMIx code, but I suspect its failures are going to be similar to the OPAL failure (OPAL is one of the layers in Open MPI): https://github.com/open-mpi/ompi/blob/1fd30debf34cac682e88ce484a5626d889fd2114/opal/mca/if/posix_ipv4/if_posix.c#L83-L87

I.e., it's literally just calling `socket(..., SOCK_DGRAM, ...)` to create a endpoint so that we can query some network layer attributes.  If that system call is failing, there's little that Open MPI (or PMIx) can do -- that's a failure in the OS/environment somewhere.  ☹️

**Comment by kpwelsh** [comment]
Not totally. But we have stopped investigating.

It seems to be an issue that we only see with WSL1. Testing it on other
devices with a native Linux OS we don't see this at all. ATM we are just
assuming it would work fine on WSL2 as well.

So I don't have any answers for why it happens, but lacking full support
for WSL1 is not that big of a deal for us.

Wish I had some more useful information :).

If no one else is interested, this can be closed.

**Comment by jsquyres** [comment]
Ok, thanks for the followup.  Since this is WSL1, let's go ahead and close this.  Someone can find it via Google and open a new ticket if they really want to (but I hope they don't 😄).
